src/scripts/modules/carousel.js

```diff
@@ -0,0 +1,92 @@
+let slideContainer;
+let nextBtn;
+let prevBtn;
+let interval;
+let slides;
+let index;
+let slideWidth;
+let firstClone;
+let lastClone;
+
+function cacheDom() {
+  slideContainer = document.querySelector(".slides");
+  nextBtn = document.querySelector(".next-btn");
+  prevBtn = document.querySelector(".prev-btn");
+  interval = 5000;
+  slides = document.querySelectorAll(".slide");
+  index = 0;
+  slideWidth = slides[index].clientWidth || window.onresize;
+  firstClone = slides[0].cloneNode(true);
+  lastClone = slides[slides.length - 1].cloneNode(true);
+}
+
+function bindEvents() {
+  slideContainer.addEventListener("transitionend", () => {
+    slides = getSlides();
+    if (slides[index] === firstClone) {
+      slideContainer.style.transition = "none";
+      index = 1; //resetting the index and the slideWidth back to the first slide
+      slideContainer.style.transform = `translateX(${-slideWidth * index}px)`;
+    }
+
+    if (slides[index] === lastClone) {
+      slideContainer.style.transition = "none";
+      index = slides.length - 2;
+      slideContainer.style.transform = `translateX(${-slideWidth * index}px)`;
+    }
+  });
+  nextBtn.addEventListener("click", moveToNextSlide);
+  prevBtn.addEventListener("click", moveToPreviousSlide);
+  // document.addEventListener("resize", resizeSlide);
+}
+
+function render() {
+  slideContainer.append(firstClone);
+  slideContainer.prepend(lastClone);
+}
+
+//slides resize on window change
+window.onresize = function () {
+  slideWidth = slides[index].clientWidth;
+  return slideWidth;
+};
+
+// function resizeSlide() {
+//   slideWidth = slides[index].clientWidth;
+//   console.log(slideWidth);
+//   return slideWidth;
+// }
+
+//includes the first and the last clone
+const getSlides = () => document.querySelectorAll(".slide");
+
+function moveToNextSlide() {
+  slides = getSlides();
+  if (index >= slides.length - 1) return;
+  index++;
+  slideContainer.style.transition = ".7s ease-out";
+  slideContainer.style.transform = `translateX(${-slideWidth * index}px)`; //the index increases so the amount amount a slide is transitioned increases
+}
+
+function moveToPreviousSlide() {
+  if (index <= 0) return;
+  index--;
+  slideContainer.style.transition = ".7s ease-out";
+  slideContainer.style.transform = `translateX(${-slideWidth * index}px)`;
+}
+
+function startSlide() {
+  setInterval(() => {
+    moveToNextSlide();
+  }, interval);
+}
+
+function init() {
+  cacheDom();
+  bindEvents();
+  render();
+  startSlide();
+}
+init();
```

Comment:
Wow, you did it all by yourself? That's awesome, congrats! 🎉 

Yes, there's always a lot of room for improvement.
For this approach the only bug I found it the `setInterval`. When the user clicks next/back, the setInterval should restart the timer. To do it, you'll need to clean the current one using `clearTimeout` and start a new one.

---

Regarding the approach itself, this works just fine! 

One that would also work without the window resize is using percentages instead of pixels using flexbox. It's tricky to explain in a short PR comment and couldn't find any simple tutorial for you. I'll give you a short overview of the final HTML markup.

```pug
.carousel // width: 50vw; height: 100vh; - it wraps everything and has overflow hidden.
  .carousel-canvas // width: 100% * nr of slides. is flexbox. this is where you apply the translate in percentage eg transform: translate(25%) (if it has 4 photos, it's the 2nd photo)
     .slide // width: 100% / nr of slides; it has flex-shrink: 0; flexbox 
        .slide-img // the image itself with object-fit: cover; to fulfill the entire area
```

Leave this for another day when you feel inspired, this one can steal you some hours :p

Reply:
Well if by 'all by yourself' you mean first going through a bunch tutorials/codepens/articles, then yes 😄 The bug I see is  when you resize the window the next slide peeks out until it's transitioned, I might have an idea of how to fix it that I need to try yet.  I definitely want to try your solution, it sounds really cool! I already tried something with flex but it didn't work out. What does ``nr`` stand for?

Reply:
>  To do it, you'll need to clean the current one using clearTimeout and start a new one.

it's not ```clearInterval```?

Reply:
I don't suppose you had this in mind?

```
function cacheDom() {        //not really the right place for this?
  startInterval = setInterval(moveToNextSlide, interval);
}
function bindEvents() {
  nextBtn.addEventListener("click", () => {
    clearInterval(startInterval);
    setInterval(moveToNextSlide, interval);
  });
}
```

Reply:
> first going through a bunch tutorials/codepens/articles, 

You'll do that a lot of times during your developer career, it's supposed to be like that! Learning by yourself ;)

> What does nr stand for?

nr = number

> it's not clearInterval?

Yes, it is, I meant that, my apologies!

Reply:
In your solution, you are almost there. The only thin missing was assigning the new interval to the same variable, so that it can be later cleared again. Like this:

```js
  nextBtn.addEventListener("click", () => {
    clearInterval(startInterval);
    startInterval = setInterval(moveToNextSlide, interval);
  });
```

Reply:
As far as I can tell the problem is still there, there's a weird delay when clicking the next button

Reply:
Hum, I tried it on my side, and it works fine. Perhaps you could [create a Loom](https://www.loom.com/), showing the problem?